### PR TITLE
[NTH-85]  Login

### PR DESCRIPTION
**Pull request by raghdarizq**
- **The `userLoginController` function handles the user login process.**

### CODE REVIEW

File: src/controllers/userLogin.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+const userLoginController = async (req: Request, res: Response, next: NextFunction) => {
+  const { email, password } = req.body;
+
+  try {
+    await userLoginSchema.validate({ email, password });
+
+    const user = await LoginByEmail(email);
+
+    if (!user) {
+      throw templateErrors.NOT_FOUND('user not found');
```
Comment (LinaYahya) [inline src/controllers/userLogin.ts:21]:
```suggestion
      throw templateErrors.BAD_REQUEST('wrong email or password');
```

File: src/controllers/userLogin.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+const userLoginController = async (req: Request, res: Response, next: NextFunction) => {
+  const { email, password } = req.body;
+
+  try {
+    await userLoginSchema.validate({ email, password });
+
+    const user = await LoginByEmail(email);
+
+    if (!user) {
```
Comment (LinaYahya) [inline src/controllers/userLogin.ts:20]: please do check user is active

File: src/controllers/userLogin.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      throw templateErrors.FORBIDDEN('No secret key found');
+    }
+    const token = jwt.sign({ email: user.email }, config.SECRET_KEY, { expiresIn: '1h' });
+
+    res.json({
+      message: 'user logged in successfully',
+      token,
+    });
+  } catch (err) {
+    next(err);
```
Comment (LinaYahya) [inline src/controllers/userLogin.ts:39]: handle validation error please

File: src/controllers/userLogin.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    const user = await LoginByEmail(email);
+
+    if (!user) {
+      throw templateErrors.NOT_FOUND('user not found');
+    }
+
+    const passwordMatch = await bcrypt.compare(password, user.password);
+
+    if (!passwordMatch) {
+      throw templateErrors.UNAUTHORIZED('The password is invalid');
```
Comment (LinaYahya) [inline src/controllers/userLogin.ts:27]:
```suggestion
      throw templateErrors.BAD_REQUEST('wrong email or password');
```

File: src/controllers/userLogin.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    if (!config.SECRET_KEY) {
+      throw templateErrors.FORBIDDEN('No secret key found');
+    }
+    const token = jwt.sign({ email: user.email }, config.SECRET_KEY, { expiresIn: '1h' });
```
Comment (LinaYahya) [inline src/controllers/userLogin.ts:32]: please use sign async instead of blocking code using synchronous methods

Comment (LinaYahya) [reply src/controllers/userLogin.ts:32]: and please follow our agreed payload to sign token [Authentication middleware](https://www.notion.so/BE-Authentication-middleware-9d8f274a43d440cfb7dbeb0e8384b741?pvs=4)

File: src/routes/userLogin.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { Router } from 'express';
+// import { RolesForSelect } from '../types';
+// import { checkAuth } from '../middlewares';
+import { userLoginController } from '../controllers';
+
+const router = Router();
+router.post('/', userLoginController);
+// router.post('/', checkAuth(RolesForSelect.therapist), userLoginController);
```
Comment (LinaYahya) [inline src/routes/userLogin.ts:8]: remove comments

File: src/routes/userLogin.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { Router } from 'express';
+// import { RolesForSelect } from '../types';
+// import { checkAuth } from '../middlewares';
+import { userLoginController } from '../controllers';
+
+const router = Router();
+router.post('/', userLoginController);
```
Comment (LinaYahya) [inline src/routes/userLogin.ts:7]:
```suggestion
router.post('/login', login);
```

File: src/routes/index.ts [hunk 8]
```diff
@@ -1,10 +1,12 @@
 import { Router } from 'express';
 import appointmentsRouter from './appointment';
 import therapistRouter from './therapist';
+import userRouter from './userLogin';
 
 const router = Router();
 
 router.use('/appointments', appointmentsRouter);
 router.use('/therapists', therapistRouter);
+router.use('/userLogin', userRouter);
```
Comment (LinaYahya) [inline src/routes/index.ts:10]:
```suggestion
router.use('/auth', userRouter);
```

File: src/controllers/userLogin.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import { NextFunction, Request, Response } from 'express';
+import bcrypt from 'bcrypt';
+import jwt from 'jsonwebtoken';
+import dotenv from 'dotenv';
+import { userLoginSchema } from '../helpers/validation';
+import { templateErrors } from '../helpers';
+import { LoginByEmail } from '../services';
+import config from '../config';
+
+dotenv.config();
```
Comment (LinaYahya) [inline src/controllers/userLogin.ts:10]:
```suggestion
```

File: src/helpers/validation/userLogin.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import * as Yup from 'yup';
+
+const userLoginSchema = Yup.object().shape({
+  email: Yup.string().email('Invalid email address').required('Email is required'),
+  password: Yup.string().required('Password is required'),
```
Comment (LinaYahya) [inline src/helpers/validation/userLogin.ts:5]: please do validate password

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by ahmedmurtaja — APPROVED** [review]

**Review by MohannadSabaa — APPROVED** [review]

**Review by YasserAmro0 — APPROVED** [review]

File: src/controllers/auth.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+    if (!passwordMatch) {
+      throw templateErrors.BAD_REQUEST('Wrong email or password');
+    }
+
+    if (user && !user.isActive) {
+      throw templateErrors.BAD_REQUEST('Your account is not activated yet. Please check your email for activation instructions.');
+    }
+
+    const payload: {
```
Comment (LinaYahya) [inline src/controllers/auth.ts:34]:
```suggestion
    const payload: IPayload
```

File: src/controllers/auth.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    if (user.role === 'therapist') {
+      payload.therapistId = user?.therapist?.id as number;
+    }
+
+    const token = await generateToken(payload);
+
+    res.json({
+      message: 'User logged in successfully',
+      token,
+      user,
```
Comment (LinaYahya) [inline src/controllers/auth.ts:52]:
```suggestion
      data: {access_token: token},
```

File: src/controllers/auth.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
+import { NextFunction, Request, Response } from 'express';
+import bcrypt from 'bcrypt';
+import * as yup from 'yup';
+import { userLoginSchema } from '../helpers/validation';
+import { templateErrors, generateToken } from '../helpers';
+import { loginByEmail } from '../services';
+import { TherapistAttributes, UsersAttributes } from '../types';
+
+interface TherapistAndUser extends UsersAttributes {
```
Comment (LinaYahya) [inline src/controllers/auth.ts:9]: please move this to types folder

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by LinaYahya — APPROVED** [review]

**Pull request merged**